### ISSUE TITLE
needless_lifetimes ignores trait bounds

### ISSUE THREAD

**Issue by Manishearth** [Manishearth/rust-clippy#292]
`fn foo<'a, T: Foo<'a>>(&'a u8, T)` will hit the lint. It shouldn't, whenever we have lifetimes in trait bounds we should bail.

**Comment by Manishearth** [comment]
https://github.com/Manishearth/rust-clippy/issues/417

**Comment by fhartwig** [comment]
I think this should be re-opened. It's not really the same problem as #417, even though they're both false positives. #417 is basically about mis-counting input and output lifetimes, while the problem here is that you're not allowed to elide lifetimes in trait bounds.

**Comment by llogiq** [comment]
I'd like to add that we still have false positives even after #599 – The followig (from rust-bio) is still linted:

    pub fn find_all<'a, I: Iterator<Item=&'a u8>>(&'a self, text: I) -> Matches<I> {
        Matches { shiftand: self, active: 0, text: text.enumerate() }
    }

**Comment by dtolnay** [comment]
Users of Serde are going to start hitting this really badly in the next release as we are adding a lifetime to the Deserializer trait.

```rust
fn f<'de, D: Deserializer<'de>>(deserializer: D) -> Result<F, D::Error> { /* ... */ }
```

It would be great to fix this before Serde 1.0.

**Comment by mcarton** [comment]
@dtolnay I believe a simple case like this is already not linted against?
https://github.com/Manishearth/rust-clippy/blob/0432f1efd31f8cd58faea180d8e44a786bc6fc4c/tests/ui/lifetimes.rs#L128-L129
https://github.com/Manishearth/rust-clippy/blob/0432f1efd31f8cd58faea180d8e44a786bc6fc4c/tests/ui/lifetimes.stderr#L88-L92

**Comment by dtolnay** [comment]
@mcarton for whatever reason that is not the same. In particular:

```rust
trait WithLifetime<'a> {}

fn trait_bound<'a, T: WithLifetime<'a>>() { unimplemented!() }
```

```
warning: explicit lifetimes given in parameter types where they could be elided
 --> src/main.rs:3:1
  |
3 | fn trait_bound<'a, T: WithLifetime<'a>>() { unimplemented!() }
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: #[warn(needless_lifetimes)] on by default
  = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_lifetimes
```
